### PR TITLE
Add host inventory metrics to ec2 metricset

### PR DESCRIPTION
**Pull request by kaiyan-sheng**
## What does this PR do?

This PR is to add proposed host common fields into `ec2` metricset:
- host.id
- host.name
- host.cpu.pct
- host.network.in.bytes
- host.network.in.packets
- host.network.out.bytes
- host.network.out.packets
- host.disk.read.bytes
- host.disk.write.bytes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Start Metricbeat AWS module: `./metricbeat modules enable aws`
2. Edit `modules.d/aws.yml` to only include `ec2` metricset:
```
- module: aws
  period: 5m
  credential_profile_name: elastic-beats
  metricsets:
    - ec2
```
3. Change add_host_metadata processor config in metricbeat.yml file:
```
processors:
  - add_host_metadata:
      replace_fields: false
```
4. Start metricbeat
5. You should see metrics from `ec2` metricset and includes fields listed above.
(`host.name` only equals to EC2 instance name when EC2 instance name is available.)

## Event Example
```
{
    "@timestamp": "2017-10-12T08:05:34.853Z",
    "aws": {
        "ec2": {
            "cpu": {
                "credit_balance": 144,
                "credit_usage": 0.05477,
                "surplus_credit_balance": 0,
                "surplus_credits_charged": 0,
                "total": {
                    "pct": 1.166484417714406
                }
            },
            "diskio": {
                "read": {
                    "bytes": 0,
                    "bytes_per_sec": 0,
                    "count": 0,
                    "count_per_sec": 0
                },
                "write": {
                    "bytes": 0,
                    "bytes_per_sec": 0,
                    "count": 0,
                    "count_per_sec": 0
                }
            },
            "instance": {
                "core": {
                    "count": 1
                },
                "image": {
                    "id": "ami-04bc3da8f14823e88"
                },
                "monitoring": {
                    "state": "disabled"
                },
                "private": {
                    "dns_name": "ip-172-31-9-119.us-west-1.compute.internal",
                    "ip": "172.31.9.119"
                },
                "public": {
                    "dns_name": "ec2-13-52-163-56.us-west-1.compute.amazonaws.com",
                    "ip": "13.52.163.56"
                },
                "state": {
                    "code": 16,
                    "name": "running"
                },
                "threads_per_core": 1
            },
            "network": {
                "in": {
                    "bytes": 6644.4,
                    "bytes_per_sec": 22.148,
                    "packets": 44.6,
                    "packets_per_sec": 0.14866666666666667
                },
                "out": {
                    "bytes": 6016.6,
                    "bytes_per_sec": 20.055333333333333,
                    "packets": 35.4,
                    "packets_per_sec": 0.118
                }
            },
            "status": {
                "check_failed": 0,
                "check_failed_instance": 0,
                "check_failed_system": 0
            }
        },
        "tags": {
            "Name": "mysql-test",
            "created-by": "ks"
        }
    },
    "cloud": {
        "account": {
            "id": "428152502467",
            "name": "elastic-beats"
        },
        "availability_zone": "us-west-1b",
        "instance": {
            "id": "i-0516ddaca5c1d231f",
            "name": "mysql-test"
        },
        "machine": {
            "type": "t2.micro"
        },
        "provider": "aws",
        "region": "us-west-1"
    },
    "event": {
        "dataset": "aws.ec2",
        "duration": 115000,
        "module": "aws"
    },
    "host": {
        "cpu": {
            "pct": 1.166484417714406
        },
        "disk": {
            "read": {
                "bytes": 0
            }
        },
        "diskio": {
            "write": {
                "bytes": 0
            }
        },
        "id": "i-0516ddaca5c1d231f",
        "name": "mysql-test",
        "network": {
            "in": {
                "bytes": 6644.4,
                "packets": 44.6
            },
            "out": {
                "bytes": 6016.6,
                "packets": 35.4
            }
        }
    },
    "metricset": {
        "name": "ec2",
        "period": 10000
    },
    "service": {
        "type": "aws"
    }
}
```

### CODE REVIEW

File: x-pack/metricbeat/module/aws/ec2/ec2.go [hunk 1]
```diff
@@ -309,6 +314,37 @@ func calculateRate(resultMetricsetFields common.MapStr, monitoringState string)
...
+	}
+
+	hostFieldTable := map[string]string{
+		"cpu.total.pct":       "host.cpu.pct",
+		"network.in.bytes":    "host.network.in.bytes",
+		"network.out.bytes":   "host.network.out.bytes",
+		"network.in.packets":  "host.network.in.packets",
+		"network.out.packets": "host.network.out.packets",
+		"diskio.read.bytes":   "host.disk.read.bytes",
+		"diskio.write.bytes":  "host.diskio.write.bytes",
```
Comment (jsoriano) [inline x-pack/metricbeat/module/aws/ec2/ec2.go:336]: Use the same field name in both:
```suggestion
		"diskio.read.bytes":   "host.disk.read.bytes",
		"diskio.write.bytes":  "host.disk.write.bytes",
```

File: x-pack/metricbeat/module/aws/ec2/_meta/data.json [hunk 2]
```diff
@@ -95,6 +95,33 @@
...
+            "write": {
+                "bytes": 0
+            }
+        },
+        "id": "i-0516ddaca5c1d231f",
+        "name": "mysql-test",
+        "network": {
+            "in": {
+                "bytes": 6644.4,
+                "packets": 44.6
```
Comment (jsoriano) [inline x-pack/metricbeat/module/aws/ec2/_meta/data.json:117]: Is this a counter? Why is it a decimal number?

File: CHANGELOG.next.asciidoc [hunk 3]
```diff
@@ -595,6 +595,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Added cache and connection_errors metrics to status metricset of MySQL module {issue}16955[16955] {pull}19844[19844]
 - Update MySQL dashboard with connection errors and cache metrics {pull}19913[19913] {issue}16955[16955]
 - Add cloud.instance.name into aws ec2 metricset. {pull}20077[20077]
+- Add host inventory metrics into aws ec2 metricset. {pull}20171[20171]
```
Comment (jsoriano) [inline CHANGELOG.next.asciidoc:598]: We will have to add the fields to some `fields.yml` file so they are documented and have a mapping.

File: x-pack/metricbeat/module/aws/ec2/_meta/data.json (hunk 2 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/aws/ec2/_meta/data.json:117]: It's a gauge actually and it's an average value among 5 minutes:

Screen Shot 2020-07-31 at 8 36 46 AM

File: CHANGELOG.next.asciidoc (hunk 3 above)
Comment (kaiyan-sheng) [reply CHANGELOG.next.asciidoc:598]: Yes I'm thinking once we are set on these fields, we can add them to ECS. I can add these fields in `fields.yml` for now and once these fields are in ECS, I assume we don't need them in `fields.yml` anymore?

Comment (kaiyan-sheng) [reply CHANGELOG.next.asciidoc:598]: I just added the host fields into fields.yml 😄 Thank you for the review!!

File: metricbeat/docs/fields.asciidoc [hunk 4]
```diff
@@ -1586,7 +1650,7 @@ type: long
 
 
 
-*`aws.cloudwatch.namespace`*::
+*`host.cloudwatch.namespace`*::
```
Comment (jsoriano) [inline metricbeat/docs/fields.asciidoc:1653]: Are these changes expected? I'd say these fields should continue under `aws`.

File: x-pack/metricbeat/module/aws/_meta/fields.yml [hunk 5]
```diff
@@ -30,3 +30,27 @@
...
+          description: The number of packets received on all network interfaces by the host in a given period of time.
+        - name: network.out.packets
+          type: scaled_float
+          description: The number of packets sent out on all network interfaces by the host in a given period of time.
+        - name: disk.read.bytes
+          type: scaled_float
+          description: The total number of bytes read successfully in a given period of time.
+        - name: disk.write.bytes
+          type: scaled_float
+          description: The total number of bytes write successfully in a given period of time.
```
Comment (jsoriano) [inline x-pack/metricbeat/module/aws/_meta/fields.yml:56]: We should move these fields to some common file, as they are supposed to be used in more places.

If it is kept here, it should be moved before the `aws` group. Metrics of metricsets are appended inside the last group of this file, this is probably causing the documented metrics to be moved from `aws` to `host`.

File: metricbeat/docs/fields.asciidoc (hunk 4 above)
Comment (kaiyan-sheng) [reply metricbeat/docs/fields.asciidoc]: Not expected 😄 Thanks for catching this! This is resolved by moving the host fields into common fields.yml in libbeat.

File: x-pack/metricbeat/module/aws/_meta/fields.yml (hunk 5 above)
Comment (kaiyan-sheng) [reply x-pack/metricbeat/module/aws/_meta/fields.yml]: Done! This is moved into `libbeat/_meta/fields.common.yml`

**Comment by sorantis** [comment]
@kaiyan-sheng, @jsoriano, in the following configuration

```
processors:
  - add_host_metadata:
      replace_fields: false
```

what does the `replace_fields` flag do?

**Comment by jsoriano** [comment]
> @kaiyan-sheng, @jsoriano, in the following configuration
> 
> ```
> processors:
>   - add_host_metadata:
>       replace_fields: false
> ```
> 
> what does the `replace_fields` flag do?

It can be used to control if `add_host_metadata` should replace existing `host` fields. It defaults to `true`, so this processor overwrites existing `host` fields, and can be set to `false` to avoid this overwrite, that is what we are going to want with inventory fields added by modules. We added this option to support inventory fields, while avoiding the introduction of a breaking change by now.

Full discussion and change to add this flag can be found here: https://github.com/elastic/beats/pull/20490

**Comment by sorantis** [comment]
ok, my expectation was that this field either map or remove the respective ec2.* fields in favor of host.* fields.
Because the resulting document contains duplicate metrics (e.g.: `ec2.cpu.total.pct` = `host.cpu.pct`), which increases the doc size by 10-11%

**Comment by kaiyan-sheng** [comment]
@sorantis Currently we are duplicating the fields so yes we have both `ec2.*` fields and `host.*` fields. But I'm planning to create a PR to remove the duplicated `ec2.*` fields and use `alias` to keep backward compatibility. For example:
```
- name: aws.ec2
  type: group
  fields:
    - name: cpu.total.pct
      type: alias
      path: host.cpu.pct
      migration: true
    - name: network.in.bytes
      type: alias
      path: host.network.in.bytes
      migration: true
```

**Comment by sorantis** [comment]
Great! Thanks @kaiyan-sheng .

**Pull request merged**